### ISSUE TITLE
[native-image] Compiled executable segfaults when trying to write to "reserved3" field of JNINativeInterface_ struct

### ISSUE THREAD

**Issue by Berstanio** [oracle/graal#6391]
**Describe the issue**
The JNINativeInterface struct has the following layout:
```c
struct JNINativeInterface_ {
    void *reserved0;
    void *reserved1;
    void *reserved2;

    void *reserved3;
    jint (JNICALL *GetVersion)(JNIEnv *env);
    [...]
```
When trying to write to a "reserved" field, the generated native image segfaults. On a normal JVM this operation works just fine.
I honestly don't really know what the purpose of those 4 fields is, I couldn't find anything documented. I also just found them, because a library used them, that I wanted to use with native-image.

**Steps to reproduce the issue**
1. git clone https://github.com/Berstanio/ReservedSegfault.git
2. sh build_mac.sh or sh build_linux.sh

For running it as jvm:
2. sh run_jvm_java_mac.sh or sh run_jvm_java_linux.sh

**Describe GraalVM and your environment:**
 - GraalVM version 22.3.1
 - JDK major version: 11.0.18+10-jvmci-22.3-b13
 - OS: macOS Ventura or debian 11 
 - Architecture: arm64 or amd64

**More details**
What is somehow interesting is, that the reproducer actually doesn't crash, when changing the ressourse regex to `-H:IncludeResources=".*.*"` (therefor including all ressources).
However, this "workaround" only works on the reproducer and not on my actual project.

Here is the log of the actual crash:
```
[ [ SubstrateSegfaultHandler caught a segfault in thread 0x0000000152f04080 ] ]
siginfo: si_signo: 10, si_code: 1, si_addr: 0x00000001033b8028

Current timestamp: 1680978112068

General purpose register values:
  R0  0x0000000152f04080 is a thread
  R1  0x0000000100284028 is an unknown value
  R2  0x0000000103788d80 points into the image heap (read-only huge)
    is an object of type [B
  R3  0x0000000103101028 points into the image heap (read-only primitives)
    is an object of type [B
  R4  0x0000000000000000 
  R5  0x0000000000000000 
  R6  0x0000000000000000 
  R7  0x0000000000000000 
  R8  0x00000001033b8010 points into the image heap (read-only relocatables)
  R9  0x0000000152f04094 points into the thread locals for thread 0x0000000152f04080
  R10 0x000000010383c02f is an unknown value
  R11 0x0000000197c52869 is an unknown value
  R12 0x000000010383c02f is an unknown value
  R13 0x000000016db5b2d8 points into the stack for thread 0x0000000152f04080
  R14 0x000000010383c02b is an unknown value
  R15 0x00000001036c8100 points into the image heap (writable references)
    is an object of type jdk.vm.ci.aarch64.AArch64$CPUFeature
  R16 0x00000002b8c37a88 is an unknown value
  R17 0x00000001f80a5318 is an unknown value
  R18 0x0000000000000000 
  R19 0x0000000102f25ca0 is an unknown value
  R20 0x0000000103901b00 points into an aligned TLAB chunk of thread 0x0000000152f04080
    is an object of type java.util.ArrayList
  R21 0x0000000102f25cc0 is an unknown value
  R22 0x000000016db5b670 points into the stack for thread 0x0000000152f04080
  R23 0x0000000197c49000 is an unknown value
  R24 0x00000001f376b340 is an unknown value
  R25 0x0000000000000000 
  R26 0x0000000000000000 
  R27 0x0000000103000000 is the heap base
  R28 0x0000000152f04080 is a thread
  R29 0x000000016db5b7c0 points into the stack for thread 0x0000000152f04080
  R30 0x000000010269f0b4 points into AOT compiled code org.example.Main.writeToReserved3(Main.java)
  SP  0x000000016db5b2c0 points into the stack for thread 0x0000000152f04080
  PC  0x000000010383bfb0 is an unknown value
  
Printing Instructions (ip=0x000000010383bfb0):
  0x000000010383bf90: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
  0x000000010383bfa0: 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x08 0x00 0x40 0xf9
  0x000000010383bfb0: 0x1f 0x0d 0x00 0xf9 0xc0 0x03 0x5f 0xd6 0x01 0x00 0x00 0x00 0x1c 0x00 0x00 0x00
  0x000000010383bfc0: 0x00 0x00 0x00 0x00 0x1c 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x1c 0x00 0x00 0x00

Top of stack (sp=0x000000016db5b2c0):
  0x000000016db5b2c0: 0x000000016db5b7c0 0x00000001023d6438
  0x000000016db5b2d0: 0x000000016db5b238 0x000000010269f098
  0x000000016db5b2e0: 0x000000016db5b2c0 0x0000000000000000
  0x000000016db5b2f0: 0x000000010383bfac 0x00000001032cfe18
  0x000000016db5b300: 0x000000016db5b7c0 0x000000010269efc8
  0x000000016db5b310: 0x00000000006c7ee0 0x00000000009010e0
  0x000000016db5b320: 0x0000000103284028 0x00000001035b96a8
  0x000000016db5b330: 0x000000016db5b7c0 0x00000001022b56b4
  0x000000016db5b340: 0x0000000100000000 0x0000000100000000
  0x000000016db5b350: 0x00000001039010e0 0x000000010235939c
  0x000000016db5b360: 0x0000000000000001 0x0000000102efc5d0
  0x000000016db5b370: 0x0000000102efc960 0x0000000100000000
  0x000000016db5b380: 0x000000016db5b7c0 0x00000001022b5308
  0x000000016db5b390: 0x0000000000000000 0x0000000000000000
  0x000000016db5b3a0: 0x0000000000000000 0x0000000102efc960
  0x000000016db5b3b0: 0x000000016db5b7e8 0x0000000197c49000
  0x000000016db5b3c0: 0x000000016db5b7c0 0x00000001022de60c
  0x000000016db5b3d0: 0x000000016db5b3f0 0x000000016db5b498
  0x000000016db5b3e0: 0x000000016db5b460 0xe912000197bfd400
  0x000000016db5b3f0: 0x000000016db5b7e8 0x0000000100000000
  0x000000016db5b400: 0x0000000000000000 0x0000000000000000
  0x000000016db5b410: 0x0000000000000000 0x0000000000000000
  0x000000016db5b420: 0x0000000000000000 0x0000000000000000
  0x000000016db5b430: 0x0000000000000000 0x0000000000000000
  0x000000016db5b440: 0x0000000000000000 0x0000000000000000
  0x000000016db5b450: 0x0000000000000000 0x0000000000000000
  0x000000016db5b460: 0x0000000000000000 0x0000000000000000
  0x000000016db5b470: 0x0000000000000000 0x0000000000000000
  0x000000016db5b480: 0x0000000000000000 0x0000000000000000
  0x000000016db5b490: 0x0000000000000000 0x0000000000000000
  0x000000016db5b4a0: 0x00000001f376b340 0x0000000197c49000
  0x000000016db5b4b0: 0x000000016db5b670 0x0000000102f25cc0

Top frame info:
  Does not look like a Java Frame. Use JavaFrameAnchors to find LastJavaSP:
  
Threads:
  0x0000000153804080 STATUS_IN_NATIVE (ALLOW_SAFEPOINT) "Reference Handler" - 0x0000000103513740, daemon, stack(0x000000016db60000,0x000000016dbe3000)
  0x0000000152f04080 STATUS_IN_JAVA (PREVENT_VM_FROM_REACHING_SAFEPOINT) "main" - 0x0000000103513690, stack(0x000000016d360000,0x000000016db5c000)
  
VM thread locals for the failing thread 0x0000000152f04080:
  0 (8 bytes): JNIThreadLocalEnvironment.jniFunctions = (bytes) 
    0x0000000152f04080: 0x00000001033b8010
  8 (8 bytes): StackOverflowCheckImpl.stackBoundaryTL = (Word) 1 (0x0000000000000001)
  16 (4 bytes): Safepoint.safepointRequested = (int) 2147458305 (0x7fff9d01)
  20 (4 bytes): StatusSupport.statusTL = (int) 1 (0x00000001)
  24 (32 bytes): ThreadLocalAllocation.regularTLAB = (bytes) 
    0x0000000152f04098: 0x0000000103900000 0x0000000103a00000
    0x0000000152f040a8: 0x0000000103905d78 0x0000000000000000
  56 (8 bytes): PlatformThreads.currentThread = (Object) java.lang.Thread (0x0000000103513690)
  64 (8 bytes): JavaFrameAnchors.lastAnchor = (Word) 6135591640 (0x000000016db5b2d8)
  72 (8 bytes): AccessControlContextStack = (Object) null
  80 (8 bytes): ExceptionUnwind.currentException = (Object) null
  88 (8 bytes): IdentityHashCodeSupport.hashCodeGeneratorTL = (Object) null
  96 (8 bytes): IsolatedCompileClient.currentClient = (Object) null
  104 (8 bytes): IsolatedCompileContext.currentContext = (Object) null
  112 (8 bytes): JNIObjectHandles.handles = (Object) com.oracle.svm.core.handles.ThreadLocalHandles (0x00000001039011a0)
  120 (8 bytes): JNIThreadLocalPendingException.pendingException = (Object) null
  128 (8 bytes): JNIThreadLocalPinnedObjects.pinnedObjectsListHead = (Object) null
  136 (8 bytes): JNIThreadOwnedMonitors.ownedMonitors = (Object) null
  144 (8 bytes): NoAllocationVerifier.openVerifiers = (Object) null
  152 (8 bytes): ThreadingSupportImpl.activeTimer = (Object) null
  160 (8 bytes): SubstrateDiagnostics.threadOnlyAttachedForCrashHandler = (bytes) 
    0x0000000152f04120: 0x0000000000000000
  168 (8 bytes): ThreadLocalAllocation.allocatedBytes = (Word) 0 (0x0000000000000000)
  176 (8 bytes): VMThreads.IsolateTL = (Word) 4345298944 (0x0000000103000000)
  184 (8 bytes): VMThreads.OSThreadHandleTL = (Word) 8374336128 (0x00000001f3263a80)
  192 (8 bytes): VMThreads.OSThreadIdTL = (Word) 8374336128 (0x00000001f3263a80)
  200 (8 bytes): VMThreads.StackBase = (Word) 6135595008 (0x000000016db5c000)
  208 (8 bytes): VMThreads.StackEnd = (Word) 6127222784 (0x000000016d360000)
  216 (8 bytes): VMThreads.StartedByCurrentIsolate = (bytes) 
    0x0000000152f04158: 0x0000000000000000
  224 (8 bytes): VMThreads.nextTL = (Word) 0 (0x0000000000000000)
  232 (8 bytes): VMThreads.unalignedIsolateThreadMemoryTL = (Word) 5686444160 (0x0000000152f04080)
  240 (4 bytes): ActionOnExitSafepointSupport.actionTL = (int) 0 (0x00000000)
  244 (4 bytes): ActionOnTransitionToJavaSupport.actionTL = (int) 0 (0x00000000)
  248 (4 bytes): ImplicitExceptions.implicitExceptionsAreFatal = (int) 0 (0x00000000)
  252 (4 bytes): StackOverflowCheckImpl.yellowZoneStateTL = (int) 2130640638 (0x7efefefe)
  256 (4 bytes): StatusSupport.safepointBehaviorTL = (int) 1 (0x00000001)
  260 (4 bytes): ThreadingSupportImpl.currentPauseDepth = (int) 0 (0x00000000)
  
No VMOperation in progress

The 15 most recent VM operation status changes (oldest first):
  
Counters:
  
Java frame anchors for the failing thread 0x0000000152f04080:
  Anchor 0x000000016db5b2d8 LastJavaSP 0x000000016db5b2c0 LastJavaIP 0x000000010269f098
  
Stacktrace for the failing thread 0x0000000152f04080:
  SP 0x000000016db5b2c0 IP 0x000000010383bfb0  IP is not within Java code. Trying frame anchor of last Java frame instead.
  SP 0x000000016db5b2c0 IP 0x000000010269f098  [image code] org.example.Main.writeToReserved3(Main.java)
  SP 0x000000016db5b310 IP 0x000000010269efc8  [image code] org.example.Main.main(Main.java:13)
  SP 0x000000016db5b340 IP 0x00000001022b56b4  [image code] com.oracle.svm.core.JavaMainWrapper.runCore0(JavaMainWrapper.java:175)
  SP 0x000000016db5b390 IP 0x00000001022b5308  [image code] com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:135)
  SP 0x000000016db5b390 IP 0x00000001022b5308  [image code] com.oracle.svm.core.JavaMainWrapper.doRun(JavaMainWrapper.java:232)
  SP 0x000000016db5b3d0 IP 0x00000001022de60c  [image code] com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:218)
  SP 0x000000016db5b3d0 IP 0x00000001022de60c  [image code] com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_3148eece06270530b6e0d4d60311411342c82698(IsolateEnterStub.java:0)
  
VM mutexes:
  mutex "mainVMOperationControlWorkQueue" is unlocked.
  mutex "thread" is unlocked.
  mutex "referencePendingList" is unlocked.
  
AOT compiled code is mapped at 0x00000001022a8000 - 0x000000010279794f

Heap settings and statistics:
  Supports isolates: true
  Heap base: 0x0000000103000000
  Object reference size: 8
  Aligned chunk size: 1048576
  Incremental collections: 0
  Complete collections: 0
  
Native image heap boundaries:
  ReadOnly Primitives: 0x0000000103101028 - 0x0000000103257e58
  ReadOnly References: 0x0000000103257e58 - 0x00000001033b4da8
  ReadOnly Relocatables: 0x00000001033b8000 - 0x000000010343b2c0
  Writable Primitives: 0x000000010343c000 - 0x00000001034c4770
  Writable References: 0x00000001034c4770 - 0x00000001036ea7c0
  Writable Huge: 0x0000000103700030 - 0x0000000103721208
  ReadOnly Huge: 0x0000000103724030 - 0x0000000103835e60

Heap:
  Young generation: 
    Eden: 
      edenSpace:
        aligned: 0/0 unaligned: 0/0
    Survivors: 
      Survivor-1 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-1 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-2 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-2 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-3 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-3 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-4 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-4 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-5 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-5 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-6 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-6 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-7 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-7 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-8 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-8 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-9 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-9 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-10 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-10 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-11 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-11 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-12 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-12 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-13 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-13 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-14 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-14 To:
        aligned: 0/0 unaligned: 0/0
      Survivor-15 From:
        aligned: 0/0 unaligned: 0/0
      Survivor-15 To:
        aligned: 0/0 unaligned: 0/0
  Old generation: 
    oldFromSpace:
      aligned: 0/0 unaligned: 0/0
    oldToSpace:
      aligned: 0/0 unaligned: 0/0
    
  Unused:
    aligned: 0/0

Segfault detected, aborting process. Use runtime option -R:-InstallSegfaultHandler if you don't want to use SubstrateSegfaultHandler.
```

**Comment by nirvdrum** [comment]
Those fields are [reserved for the VM](https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#interface-function-table). Writing to them is playing with fire and I think you're seeing first-hand why. What library is writing to those fields and for what purpose?

**Comment by Berstanio** [comment]
> Those fields are [reserved for the VM](https://docs.oracle.com/en/java/javase/17/docs/specs/jni/functions.html#interface-function-table). Writing to them is playing with fire and I think you're seeing first-hand why. What library is writing to those fields and for what purpose?

The library using this technique is lwjgl3, a  game library for java game development. For the reasons of the design decisions, please take a look here: https://github.com/LWJGL/lwjgl3/issues/875 and here: https://github.com/LWJGL/lwjgl3/blob/4f9767cd5eb87c0e682ec5107b3601bc409d93a9/modules/lwjgl/core/src/main/java/org/lwjgl/system/ThreadLocalUtil.java#L17-L88

Also, I might be wrong on this, it is a while ago that I looked into the truffle source, but doesn't truffle do something similar here, to store/get the nespresso env?
https://github.com/oracle/graal/blob/5bf5d2f66818f0969f4970d77ab42eb4e67d501a/espresso/src/com.oracle.truffle.espresso.native/src/nespresso.c#L298

**Comment by thejudge156** [comment]
Hey Berstanio, 

Thanks for looking into this. I use lwjgl and was encountering this crash and could never figure out why. What is worth noting is that compressing the executable with UPX then it runs fine. However this is not a perfect solution as jars with lots of resources will be too big for UPX to handle, (ex: Minecraft).

**Comment by Berstanio** [comment]
Hey @thejudge156 ,
If you are looking for a proper workaround for the moment, I created two substitutions classes, that should fix it:
https://github.com/Berstanio/gdx-graalhelper/blob/dev/backends/lwjgl3/src/main/java/org/lwjgl/opengl/Lwjgl3NativeSubstitutions.java
https://github.com/Berstanio/gdx-graalhelper/blob/dev/backends/lwjgl3/src/main/java/com/anyicomplex/gdx/lwjgl3/svm/Lwjgl3Substitutions.java
Please note, that this two substitutions require lwjgl 3.3.2. 
Also feel free to look around in the rest of the lwjgl3 part of the repo for other lwjgl3 specific JNI/Reflection registrations. However, the project was made for another library that utilizes lwjgl3, so not all registrations apply for a plain lwjgl3 project.

**Comment by thejudge156** [comment]
Already found them, thanks though :D

**Comment by ACGNnsj** [comment]
> Hey @thejudge156 , If you are looking for a proper workaround for the moment, I created two substitutions classes, that should fix it: https://github.com/Berstanio/gdx-graalhelper/blob/dev/backends/lwjgl3/src/main/java/org/lwjgl/opengl/Lwjgl3NativeSubstitutions.java https://github.com/Berstanio/gdx-graalhelper/blob/dev/backends/lwjgl3/src/main/java/com/anyicomplex/gdx/lwjgl3/svm/Lwjgl3Substitutions.java Please note, that this two substitutions require lwjgl 3.3.2.

Thanks a lot. But I got this when building native image.
```java
Error: Classes that should be initialized at run time got initialized during image building:
 com.badlogic.gdx.utils.SharedLibraryLoader was unintentionally initialized at build time. com.badlogic.gdx.scenes.scene2d.utils.UIUtils caused initialization of this class with the following trace:
        at com.badlogic.gdx.utils.SharedLibraryLoader.<clinit>(SharedLibraryLoader.java:39)
        at com.badlogic.gdx.scenes.scene2d.utils.UIUtils.<clinit>(UIUtils.java:13)


com.oracle.svm.core.util.UserError$UserException: Classes that should be initialized at run time got initialized during image building:
 com.badlogic.gdx.utils.SharedLibraryLoader was unintentionally initialized at build time. com.badlogic.gdx.scenes.scene2d.utils.UIUtils caused initialization of this class with the following trace:
        at com.badlogic.gdx.utils.SharedLibraryLoader.<clinit>(SharedLibraryLoader.java:39)
        at com.badlogic.gdx.scenes.scene2d.utils.UIUtils.<clinit>(UIUtils.java:13)


        at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:73)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.ProvenSafeClassInitializationSupport.checkDelayedInitialization(ProvenSafeClassInitializationSupport.java:273)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.classinitialization.ClassInitializationFeature.duringAnalysis(ClassInitializationFeature.java:164)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$10(NativeImageGenerator.java:748)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:85)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$11(NativeImageGenerator.java:748)
        at org.graalvm.nativeimage.pointsto/com.oracle.graal.pointsto.AbstractAnalysisEngine.runAnalysis(AbstractAnalysisEngine.java:162)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:745)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:578)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:535)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:403)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:580)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
```
The commands are ```java -agentlib:native-image-agent=config-output-dir=G:\idea\Unciv\desktop\build\libs\config -jar "G:\idea\Unciv\desktop\build\libs\Unciv.jar" com.unciv.app.desktop.DesktopLauncher``` and ```native-image --no-fallback --report-unsupported-elements-at-runtime --trace-class-initialization=com.badlogic.gdx.utils.SharedLibraryLoader -H:+ReportExceptionStackTraces -H:IncludeResources=".*.*" -H:MaxDuplicationFactor=4.0 -H:ConfigurationFileDirectories=G:\idea\Unciv\desktop\build\libs\config -H:Path=G:\idea\Unciv\desktop\build\libs\game -jar "G:\idea\Unciv\desktop\build\libs\Unciv.jar"```.

I imported the dependencies in this way.
```kotlin
"implementation"("com.github.berstanio:gdx-graalhelper:master-SNAPSHOT"){
    exclude(group = "com.github.berstanio.gdx-graalhelper", module = "gdx-svmhelper-backend-moe")
}
```

**Comment by thejudge156** [comment]
LWJGL implemented a fix for this in their 3.3.3 snapshot

**Comment by peter-hofer** [comment]
Closing because this is not a bug and LWJGL has a fix/workaround now.
